### ISSUE TITLE
Separate Build-depends: from Suggests: in DESCRIPTION files

### ISSUE THREAD

**Issue by hughjonesd** [HenrikBengtsson/Wishlist-for-R#142]
"Suggests:" packages are guaranteed to be present when CRAN runs R CMD CHECK. So, Suggests: tends to be full of packages which are used for testing, vignette building, etc - e.g. roxygen2/testthat/withr/rmarkdown. That is confusing to non-expert users, who may not be able to tell whether a package is truly relevant for them.

If there were an optional Build-depends: list of packages, CRAN could guarantee those, and the Suggests: package could be reserved for packages that enhance functionality.

I think this could be introduced without breaking anything.

**Comment by Enchufa2** [comment]
I strongly disagree. Packages used for testing, vignette building, etc., are not build-dependencies. They are _development-dependencies_, if you like, so they are not relevant for users; and tests, vignettes... using them should be properly guarded if those packages are not present. Users don't need to run the tests or rebuild the vignettes to install a package and use it. Only packages required for installation (i.e. running `R CMD INSTALL`, which effectively _builds_ the source package and installs it) should be considered build-dependencies. Adding development-dependencies as build-dependencies would result in dependency loops and would make the life of downstream packagers much much more difficult.

**Comment by hughjonesd** [comment]
Perhaps I'm using the wrong language, and "Develop-depends:" would be a better name. I wasn't envisaging that these packages would be required every time someone builds a package - only when CRAN does its checks. Or maybe, as now, CRAN should require both Suggests: and Develop-depends:. If so, splitting the two kinds of packages would still help users figure out which packages are relevant to them, which is the main purpose.

**Comment by Enchufa2** [comment]
I would argue that _all_ Suggests already are relevant development dependencies, because they are (conditionally) used somewhere: a example, a test, a vignette... So my point is that the current model already has this purpose: users know what packages are relevant to them (Depends, Imports, LinkingTo), and developers too (previous ones + Suggests). Or could you provide any example in which this is not the case?

IMHO, the one thing that could be interesting would be a special field to clearly identify the test suite. In the same way we have `VignetteBuilder: knitr`, we could have something like `TestingFramework: testthat`.

**Comment by hughjonesd** [comment]
Consider [dplyr](https://cran.r-project.org/web/packages/dplyr/index.html). This suggests about 20 packages. Some, like microbenchmark and testthat, are presumably for testing or vignettes. Others, like dbplyr and tidyr, are plausibly complementary packages which provide related functionality for a user. A non-expert user has no way of knowing which is which.

For a more minor package, my [huxtable](https://cran.r-project.org/web/packages/huxtable/index.html) also has about 20 suggested packages. Of these, bookdown, broom, crayon, flextable, ftExtra, knitr, officer, openxlsx, rmarkdown and tinytex provide useful additional functionality, or are a common context for using huxtable. The others are just for testing and vignettes. Again, the user has no way of distinguishing the two groups.

**Comment by Enchufa2** [comment]
If a particular package provides useful additional functionality, this should be mentioned in the documentation, because there the developer has space and the means to explain why and how this package is useful. A non-expert user has no way of knowing that just by looking at a list of packages, so they would just install everything, even if only, say, 2 are really interesting for their use case.

**Comment by hughjonesd** [comment]
Sure, but if there were a list of suggested packages for the user, then they could click through and look at them. It would also help potential users make the judgment "is this relevant to me, i.e. is it related to other packages I may know and use?" 

The DESCRIPTION file is a starting point for package documentation (e.g. the package description & list of vignettes). It makes sense for it to convey some basic information to potential users.

**Comment by hughjonesd** [comment]
I don't think that works, because Enhances is already (supposed to be) used to say "this package helps these others", rather than "these other packages help this one". It's also little used and understood. (E.g. r-pkgs.org: "... it’s hard to define what that means, so we don’t recommend using Enhances.")

**Comment by JosiahParry** [comment]
CRAN already does occasional checks that ignore Suggests packages. This is referred to as "hard". I've been bitten by it once. 

Suggests can be used to have conditional functionality in a package. For example I have functions in a package that are not part of the core of the package. If you want to use one of them, it runs rlang::check_installed() to verify the suggested package is available. If not, the function is exited. It's a way of preventing unnecessary bloat for things that can be thought of as "opt in."  

There are no feature flags in R like there are in Rust for example   So using Suggests is one way of doing so.
